From 4ddb5dc0de1bd25e52627f520a49063d94f79377 Mon Sep 17 00:00:00 2001
From: Sebastian Wick <sebastian.wick@redhat.com>
Date: Wed, 29 Jul 2026 21:10:32 +0200
Subject: [PATCH] revokefs: Avoid symlink path traversal in symlink()

Use glnx_chaseat to ensure that the attacker-controlled symlink name
ends up inside the basefd directory. Note that the symlink *target* is
also attacker-controlled, but it's OK for them to be able to set any
target of their choice: that can't immediately cause traversal outside
the base directory.

[smcv: Separated from a larger commit for better reviewability]
Co-authored-by: Simon McVittie <smcv@collabora.com>
Helps: https://github.com/flatpak/flatpak/security/advisories/GHSA-qrwq-7qwx-q9rp

Backport of commit 4ddb5dc0de1bd25e52627f520a49063d94f79377.

Adapted for 1.12.9: hunk offsets for 1.12.9 writer.c; no semantic change.

---
 revokefs/writer.c | 42 ++++++++++++++++++++++--------------------
 1 file changed, 22 insertions(+), 20 deletions(-)

diff --git a/revokefs/writer.c b/revokefs/writer.c
index ad5c11a..dd6564b 100644
--- a/revokefs/writer.c
+++ b/revokefs/writer.c
@@ -268,6 +268,23 @@ chase_request_2path (RevokefsRequest *request,
                            data_size - request->arg1, out_name2);
 }
 
+static void
+chase_request_any_and_path (RevokefsRequest *request,
+                            gsize data_size,
+                            char **out_any_path,
+                            int *out_fd, char **out_name)
+{
+  if (request->arg1 >= data_size)
+    {
+      g_printerr ("Invalid path1 size\n");
+      exit (1);
+    }
+
+  *out_any_path = g_strndup ((const char *) request->data, request->arg1);
+  *out_fd = chase_parent (request->data + request->arg1,
+                          data_size - request->arg1, out_name);
+}
+
 static gboolean
 validate_path (char *path)
 {
@@ -315,22 +332,6 @@ mask_mode (int mode)
   return mode & ~S_ISUID & ~S_ISGID & ~(S_IWGRP | S_IWOTH);
 }
 
-static void
-get_any_path_and_valid_path (RevokefsRequest *request,
-                             gsize data_size,
-                             char **any_path1,
-                             char **valid_path2)
-{
-  if (request->arg1 >= data_size)
-    {
-      g_printerr ("Invalid path1 size\n");
-      exit (1);
-    }
-
-  *any_path1 = g_strndup ((const char *) request->data, request->arg1);
-  *valid_path2 = get_valid_path (request->data + request->arg1, data_size - request->arg1);
-}
-
 static ssize_t
 handle_mkdir (RevokefsRequest *request,
               gsize data_size,
@@ -404,12 +405,13 @@ handle_symlink (RevokefsRequest *request,
                 RevokefsResponse *response)
 {
   g_autofree char *from = NULL;
-  g_autofree char *to = NULL;
+  g_autofree char *to_name = NULL;
+  glnx_autofd int to_parent_fd = -1;
 
-  /* from doesn't have to be a valid path, it can be absolute or whatever */
-  get_any_path_and_valid_path (request, data_size,  &from, &to);
+  chase_request_any_and_path (request, data_size,
+                              &from, &to_parent_fd, &to_name);
 
-  if (symlinkat (from, basefd, to) == -1)
+  if (symlinkat (from, to_parent_fd, to_name) == -1)
     response->result = -errno;
   else
     response->result = 0;
-- 
2.55.0

